contracts/cosmwasm-vm/cw-cluster-connection/src/contract.rs

```diff
@@ -0,0 +1,302 @@
+            return Err(ContractError::DuplicateMessage);
+        }
+        self.store_receipt(deps.storage, src_network.clone(), conn_sn)?;
+
+        let xcall_submessage =
+            self.call_xcall_handle_message(deps.storage, &src_network, vec_msg)?;
+
+        Ok(Response::new().add_submessage(xcall_submessage))
+    }
+
+    pub fn claim_fees(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+    ) -> Result<Response, ContractError> {
+        self.ensure_relayer(deps.storage, info.sender)?;
+        let contract_balance = self.get_balance(&deps, env, self.get_denom(deps.storage));
+        let msg = BankMsg::Send {
+            to_address: self.get_relayer(deps.storage)?.to_string(),
+            amount: coins(contract_balance, self.get_denom(deps.storage)),
+        };
+        Ok(Response::new()
+            .add_attribute("action", "claim fees")
+            .add_message(msg))
+    }
+
+    pub fn set_fee(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        network_id: NetId,
+        message_fee: u128,
+        response_fee: u128,
+    ) -> Result<Response, ContractError> {
+        self.ensure_relayer(deps.storage, info.sender)?;
+        self.store_fee(deps.storage, network_id, message_fee, response_fee)?;
+        Ok(Response::new().add_attribute("action", "set_fee"))
+    }
+
+    pub fn get_fee(
+        &self,
+        store: &dyn Storage,
+        network_id: NetId,
+        response: bool,
+    ) -> Result<Uint128, ContractError> {
+        let mut fee = self.get_message_fee(store, network_id.clone());
+        if response {
+            fee += self.get_response_fee(store, network_id);
+        }
+        Ok(fee.into())
+    }
+
+    fn xcall_handle_message_reply(
+        &self,
+        _deps: DepsMut,
+        message: Reply,
+    ) -> Result<Response, ContractError> {
+        match message.result {
+            SubMsgResult::Ok(_) => Ok(Response::new()
+                .add_attribute("action", "call_message")
+                .add_attribute("method", "xcall_handle_message_reply")),
+            SubMsgResult::Err(error) => Err(ContractError::ReplyError {
+                code: message.id,
+                msg: error,
+            }),
+        }
+    }
+
+    fn xcall_handle_error_reply(
+        &self,
+        _deps: DepsMut,
+        message: Reply,
+    ) -> Result<Response, ContractError> {
+        match message.result {
+            SubMsgResult::Ok(_) => Ok(Response::new()
+                .add_attribute("action", "call_message")
+                .add_attribute("method", "xcall_handle_error_reply")),
+            SubMsgResult::Err(error) => Err(ContractError::ReplyError {
+                code: message.id,
+                msg: error,
+            }),
+        }
+    }
+
+    pub fn reply(&self, deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
```

Comment:
i dont think we would need to listen to replies in connection. if it fails on xcall everything will be reverted anyway and we are not doing anything on error reply as well.

Reply:
yeah